### ISSUE TITLE
downloadText data integrity

### ISSUE THREAD

**Issue by bfarago** [AkosUzonyi/BpCarsharing#1]
Hi Akos,
Let me think on the sporadic behaviour of the element numbers...
I am wondering if in.readLine()  in source Utils.java line 24 could return null in middle of the stream, when connection stalled due high server/network load in certain cases.

-See the official android examples, there are BufferedInputStream() only shown.
https://developer.android.com/reference/java/net/HttpURLConnection
Yep, In case of file input, the unbuffered stream is definitely faster. But some reason they use Buffered streams. I don't know. maybe there are some other things to consider...
-I think, the BufferedReader will not always blocks readLine, when unbuffered stream stalling before the connection finished or timeout occured. Perhaps, there are points when connection hangs right before a new line started (in example) but your BufferedReader's buffer could be empty momentarily. This may depends on MTU, TCP window size, buffer size on producer side, and your App thread timings on consumer side, and of course the BufferedReader's buffer size is important as well. I guess, the default buffer size is 8k, but your response now around 52k long. That means, there will be chunks anyway between producer and consumer side. So, when an end-of-chunk occured exactly at a new line, then readline should returns null because of the consumer could read the buffer til its temporary end, which is a problem. 
Probably we have to check somehow, if the connection or stream could provide this missing info about the response buffer already produced totally or not.

-There's no checks for the integrity of the processed stream. Even if I'm wrong with the thought above, and there's no early break from the processing loop, it would be a nice to have doing some integrity check here.
The minimal thing is to check if JSONArray(String) throws an exeption in the case above. Because it is also needed to get lesser elements. Let say the missing "]" does not cause json exception.

-Handle error cases, for debugging and for robustness. (getResponseCode, getErrorStream, etc)
I hope I help a lil bit.
Have a nice day!

**Comment by AkosUzonyi** [comment]
Thanks for taking the time looking at my code, and trying to help, but I do not agree with most of your points.

I think you misunderstand BufferedReader. It's readline method reads a full line, even if it is longer than the buffer size. Also it returns null if and only if the wrapped stream reached EOF. The buffer size does not affect the behavior, just increases the performance. If it worked the way you described, it would do more harm than good.

That's right though, that I don't need to read the stream line by line, I could read fixed length chunks. It might be a bit more efficient, but that's negligible.

The error handling is a good point, it would be nice to show an error dialog to the user if some error happens. Probably I'll implement it if I have some time.

**Comment by bfarago** [comment]
I Agree. That's right, there is a fill(), which blocks readLine(). I checked the source of BufferedReader().

_(Man in the middle attack is not a real case in this application, so any real integrity check may be a design overkill here, I agree as well. But server or connection could be down, timeout may happen, etc...)_

Personally, I dislike popup or modal dialogs relating to errors. A small status icon or an error id somewhere in the UI would be more than perfect, I believe.
Thanks for your work, what have you done already.
